truffleruby/src/main/ruby/post-boot/post-boot.rb

```diff
@@ -16,6 +16,27 @@
     require 'rubygems'
   rescue LoadError
   else
+    unless Gem.dir.include?(Truffle::Boot.ruby_home)
+      bad_gem_home = false
+
+      # rbenv does not set GEM_HOME
+      # rbenv-gemset has to be installed which does set GEM_HOME, it's in the subdir of Truffle::Boot.ruby_home
+      # rbenv/versions/<ruby>/gemsets
+      bad_gem_home ||= Gem.dir.include?('rbenv/versions') && !Gem.dir.include?('rbenv/versions/truffleruby')
+
+      # rvm stores gems at .rvm/gems/<ruby>@<gemset-name>
+      bad_gem_home ||= Gem.dir.include?('rvm/gems') && !Gem.dir.include?('rvm/gems/truffleruby')
+
+      # chruby stores gem in ~/.gem/<ruby>/<version>
+      bad_gem_home ||= Gem.dir.include?('.gem') && !Gem.dir.include?('.gems/truffleruby')
```

Comment:
Yeah agreed, I am already working on the RUBY_ENGINE fix. Looks promising it should not take too long.